File: src/io/ziheng/list/leetcode/ReverseLinkedList.java
package io.ziheng.list.leetcode;

import io.ziheng.list.leetcode.ListNode;
import jdk.nashorn.internal.ir.WhileNode;

/**
 * LeetCode 206. Reverse Linked List
 * https://leetcode.com/problems/reverse-linked-list/
 */
public class ReverseLinkedList {

    public ListNode reverseListRecursively(ListNode head, ListNode node) {
        if (head == null) {
            return node;
        }
        ListNode next = head.next;
        head.next = node;
        return reverseListRecursively(next, head);
    }

    public ListNode reverseListIteratively(ListNode head) {
        ListNode previousNode = null;
        ListNode currentNode = head;
        ListNode nextNode = head.next;
        while (currentNode != null) {
            nextNode = currentNode.next;
            currentNode.next = previousNode;
            previousNode = currentNode;
            currentNode = nextNode;
        }
        return previousNode;
    }
    /**
     * 反转单链表。
     *
     * 时间复杂度：O(n)
     * 空间复杂度：O(1)
     */
    public ListNode reverseList(ListNode head) {
        return reverseListRecursively(head, null);
    }

    public ListNode buildLinkedList() {
        ListNode head = new ListNode(-1);
        ListNode currentNode = head;
        for (int i = 0; i < 10; i++) {
            ListNode newNode = new ListNode(i);
            currentNode.next = newNode;
            currentNode = currentNode.next;
        }
        return head;
    }
    public String listToString(ListNode head) {
        if (head == null) {
            return "null";
        }
        StringBuilder stringBuilder = new StringBuilder();
        stringBuilder.append('[');
        while (head != null) {
            stringBuilder.append(head.val);
            stringBuilder.append(", ");
            head = head.next;
        }
        stringBuilder.append(']');
        return stringBuilder.toString();
    }
    public static void main(String[] args) {
        ReverseLinkedList reverseLinkedList = new ReverseLinkedList();
        System.out.println(
            reverseLinkedList.listToString(reverseLinkedList.buildLinkedList())
        );
        System.out.println(
            reverseLinkedList.listToString(
                reverseLinkedList.reverseList(
                    reverseLinkedList.buildLinkedList()
                )
            )
        );
    }
}
/* EOF */

File: src/cc/linkedlist/ReverseLinkedList.cc
#include <stdio.h>

/**
 * Definition for singly-linked list.
 */
class ListNode {
public:
    int val;
    ListNode *next;
    ListNode(): val(0), next(NULL) {
        // ...
    }
    ListNode(int x): val(x), next(NULL) {
        // ...
    }
    ListNode(int x, ListNode *next): val(x), next(next) {
        // ...
    }
};

/**
 * LeetCode 206. Reverse Linked List
 * https://leetcode.com/problems/reverse-linked-list/
 */
class Solution {
public:
    ListNode* reverseList(ListNode *head) {
        // return reverseListIteratively(head);
        return reverseListRecursively(head, NULL);
    }
private:
    /**
     * 反转单链表 -> 迭代
     *
     * @param headNode
     * @return ListNode*
     */
    ListNode* reverseListIteratively(ListNode *head) {
        ListNode *newHead = NULL;
        while (head) {
            ListNode *next = head -> next;
            head -> next = newHead;
            newHead = head;
            head = next;
        }
        return newHead;
    }
    /**
     * 反转单链表 -> 递归
     *
     * @param headNode
     * @param prevNode
     * @return ListNode*
     */
    ListNode* reverseListRecursively(ListNode *headNode, ListNode *prevNode) {
        if (headNode == NULL) {
            return prevNode;
        }
        ListNode* nextNode = headNode -> next;
        headNode -> next = prevNode;
        return reverseListRecursively(nextNode, headNode);
    }
};

int main(int argc, char const *argv[]) {
    ListNode listNode = ListNode(1, NULL);
    printf("%d\n", listNode.val);
    return 0;
}
/* EOF */